docs/source/cheat_sheet.rst

```diff
@@ -11,121 +11,158 @@ language represents various common types in Python 2.
+   x = ["string", u"unicode"] # type: List[six.text_type]
 
-   # If you don't know the type of something, you can use Any
-   x = mystery_function() # type: Any
-   # And if you want to have something not be type-checked, you can
-   # use ignore to suppress mypy warnings for a given line
-   # Ideally, one would never use this
-   x = confusing_function() # type: ignore
+   # Use Optional for values that could be None.
+   input_str = f() # type: Optional[str]
+   if input_str is not None:
+      print input_str
 
-   # This is how you annotate a function definition
+
+Functions
+*********
+
+.. code-block:: python
+
+   from typing import Callable, Iterable
+
+   # This is how you annotate a function definition.
    def stringify(num):
        # type: (int) -> str
        """Your function docstring goes here after the type definition."""
        return str(num)
 
-   # And here's how you specify multiple arguments
+   # And here's how you specify multiple arguments.
    def plus(num1, num2):
        # type: (int, int) -> int
        return num1 + num2
 
-   # Add type annotations for kwargs as though they were positional args
+   # Add type annotations for kwargs as though they were positional args.
    def f(num1, my_float=3.5):
        # type: (int, float) -> float
        return num1 + my_float
-   # This is how you annotate a function value
+
+   # This is how you annotate a function value.
    x = f # type: Callable[[int, float], float]
 
-   # Use Optional[Type] for objects that could be None
-   def f(input_str=None):
-       # type: (Optional[str]) -> int
-       if input_str is not None:
-           return len(input_str)
-       return 0
-
-   from typing import Mapping, MutableMapping
-   # Dict is a python dictionary
-   # MutableMapping is an abstract base class for a dict-type thing
-   # Mapping is an abtract base class for a dict-type thing that may
-   # not support writing to the mapping
+   # A generator function that yields ints is secretly just a function that
+   # returns an iterable (see below) of ints, so that's how we annotate it.
+   def f(n):
+       # type: (int) -> Iterable[int]
+       i = 0
+       while i < n:
+           yield i
+           i += 1
+
+
+What to do when puzzled
+***********************
+
+.. code-block:: python
+
+   from typing import Union, Any
+
+   # To find out what type mypy infers for an expression anywhere in
+   # your program, wrap it in reveal_type.  Mypy will print an error
+   # message with the type; remove it again before running the code.
+   reveal_type(1) # -> error: Revealed type is 'builtins.int'
+
+   # Use Union when something could be one of a few types.
+   x = [3, 5, "test", "fun"] # type: List[Union[int, str]]
+
+   # Use Any if you don't know the type of something.
+   x = mystery_function() # type: Any
+
+   # Use `ignore` if you want to have something not be type-checked,
+   # to suppress mypy warnings for a given line.
```

Comment:
IMO that's not strong enough. `ignore` is usually for suppressing complaints about things that confuse mypy or are outright bugs in mypy.

I think you need cast() in this section too.

It would also be nice to have something here about the situation where mypy complaints "Need type annotation for variable" (assignment to empty containers or None without a simple enough clue following, and None), and what to do about it.